### DISCUSSION TITLE
Will Skywalking using OceanBase as the backend storage database?

### DISCUSSION THREAD

**Discussion by longdafeng** [apache/skywalking#10507] [Ideas]
Right now, I have checked the skywalking architecture,  I found Skywalking using the MySQL as the backend storage. 

So may I ask whether Skywalking use OceanBase as the backend storage database?

In last year, we have do much more tests and found that Skywalking works very well without much change except change jdbc driver as MYSQL-JDBC 5.1.47 when use OceanBase as the backend storage database. 

What skywalking can get?
1. provide higher Availability, OceanBase has been used in much more finacial critical environment, it has been approved that it can provide "Continuous Availability", 
    1. OceanBase Database adopts the Paxos protocol and can be deployed on normal cost-effective servers. Data is written to more than half of the servers (>=3). Therefore, no data will be lost when a minority of servers fail, ensuring a recovery point objective (RPO) of 0. If the primary database fails, the remaining well-functioning servers can automatically elect a new primary database to continue service provisioning within a short period without the need for external tools or manual intervention, ensuring a recovery time objective (RTO) of less than 30 seconds. 
    2. There are more than 400+ customer who are using OceanBase, please refer to https://www.oceanbase.com/customer/home
 
2. improve performance. 
   1. If Skywalking wants to watch more than thousands of applications or machines, it is likely to meet the bottleneck of MySQL, MySQL is hard to do scale up. OceanBase is one distributed database,  it is easy to be scaled up. when performance isn't enough, just add machines, In TPC-C benchmark, there are more than 1,500 nodes working together,  at 707 million tmpC. 
   2. user can do further analysis basing on the backend storage without copy the data to other place. OceanBase has the TPC-H record, at 15.26 million QphH @30000GB.
   
3. reduce cost
  1. The storage usage of OceanBase is one third of MySQL
  2. Maybe some analysis job can move from es to OceanBase. 

4. Marketing cooperation

Other things
1. OceanBase is highly compatible with MySQL, which ensures that zero or few modification is needed for migration. In fact, we have do test basing on Skywalking with OceanBase,  we can provide the testing document if Skywalking wants. 
2. OceanBase has been work as meta database for a lot of system such as Nacos, XXL-Jobs, Juicefs, Dophinschedule and so on.  Maybe you can refer to the link https://github.com/oceanbase/oceanbase/issues/1301#issuecomment-1450044825,

**Comment by LinuxSuRen** [comment]
I'm not an expert on DB. But as you mentioned, `OceanBase is highly compatible with MySQL`. So, why do you still want to have a direct support with OB?

**Comment by tisonkun** [comment]
Since you're one of the maintainers of Oceanbase, it should be reasonable that you come up with a proposal or a prototype. Then,

1. With a technical proposal (with background), we can investigate if we'd like to do it and if there are volunteers committing to the proposal (otherwise, it's just talking).
2. With a prototype, SkyWalking users can give it a try. If the feedback is positive, it's possible to be bundled in the upstream.

Currently, this post reads to me that "**I think** integrating OB with SW is a good idea, just waiting for _some_ developers figuring out how to do it and working on it". The details and the developers committing to the suggestion are absent. If you want to record a suggestion, this should be fine.

DISCLAIMER: I'm not a PMC member of SkyWalking; just sharing some common patterns among ASF projects.

**Reply by tisonkun** [reply]
For the technical part, SW already supports MySQL as oap server storage. As long as the feature set in under the compatibility scope, you should directly use OB as a drop-in replacement. If there're compatibility issues or performance issues, coming up with concrete reports help the community know why we need a _native_ support.

**Comment by wu-sheng** [comment]
SkyWalking has MySQL-style support for years. TiDB as a MySQL mock mutant and AWS Aurora as another mutant are transparent from a client perspective.
So, what exactly are you proposing? Should I expect some unsupported scenarios?

The world has dozens of MySQL mutant/mock, we don't test them all, because it isn't worth to do.

**Comment by wu-sheng** [comment]
One more update, we have a JDBC storage enhancement, https://github.com/apache/skywalking/issues/10366
